File: boruvka_serial.py
import sys 
class Graph:
    def __init__(self, num_of_nodes):
        self.m_v = num_of_nodes
        self.m_edges = []
        self.m_component = {}

    def find_component(self, u):
        if self.m_component[u] == u:
            return u
        return self.find_component(self.m_component[u])

    def set_component(self, u):
        if self.m_component[u] == u:
            return
        else:
            for k in self.m_component.keys():
                self.m_component[k] = self.find_component(k)
    def union(self, component_size, u, v):
        if component_size[u] <= component_size[v]:
            self.m_component[u] = v
            component_size[v] += component_size[u]
            self.set_component(u)

        elif component_size[u] >= component_size[v]:
            self.m_component[v] = self.find_component(u)
            component_size[u] += component_size[v]
            self.set_component(v)

        # print(self.m_component)

    def boruvka(self):
        component_size = []
        mst_weight = 0

        minimum_weight_edge = [-1] * self.m_v

        for node in range(self.m_v):
            self.m_component.update({node: node})
            component_size.append(1)

        num_of_components = self.m_v

        while num_of_components > 2:
            for i in range(len(self.m_edges)):

                u = self.m_edges[i][0]
                v = self.m_edges[i][1]
                w = self.m_edges[i][2]

                u_component = self.m_component[u]
                v_component = self.m_component[v]

                if u_component != v_component:
                    if minimum_weight_edge[u_component] == -1 or \
                            minimum_weight_edge[u_component][2] > w:
                        minimum_weight_edge[u_component] = [u, v, w]
                    if minimum_weight_edge[v_component] == -1 or \
                            minimum_weight_edge[v_component][2] > w:
                        minimum_weight_edge[v_component] = [u, v, w]

            for node in range(self.m_v):
                if minimum_weight_edge[node] != -1:
                    u = minimum_weight_edge[node][0]
                    v = minimum_weight_edge[node][1]
                    w = minimum_weight_edge[node][2]

                    u_component = self.m_component[u]
                    v_component = self.m_component[v]

                    if u_component != v_component:
                        mst_weight += w
                        self.union(component_size, u_component, v_component)
                        # print("Added edge [" + str(u) + " - "
                                # + str(v) + "]\n"
                                # + "Added weight: " + str(w) + "\n")
                        num_of_components -= 1

            minimum_weight_edge = [-1] * self.m_v
            # print(self.m_component)
        # print("----------------------------------")
        print("Serial MST weight: " + str(mst_weight))

adjacency_list = {}

def read_graph(file_path):
    with open(file_path, 'r') as f:
        for line in f:
            if line.startswith('%'):
                continue
            u, v, w = map(int, line.split())
            if u not in adjacency_list:
                adjacency_list[u] = []
            if v not in adjacency_list:
                adjacency_list[v] = []
            adjacency_list[u].append((v, w))
            adjacency_list[v].append((u, w))

if __name__ == '__main__':

    if len(sys.argv) < 2:
        print("Usage: python main.py <mtx_file>")
        sys.exit(1)

    read_graph(sys.argv[1])
    graph = Graph(len(adjacency_list)+1)
    graph.m_edges = [(u, v, w) for u in adjacency_list for v, w in adjacency_list[u]]
    graph.boruvka()
    # print(graph.m_component)
    # print(graph.m_edges)
    # print(adjacency_list)

File: go/boruvka_serial.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type IterationDetail struct {
	IterationNumber   int     `json:"iteration_number"`
	MapTime           float64 `json:"map_time"`
	ReduceTime        float64 `json:"reduce_time"`
	PreComponentCount int     `json:"pre_component_count"`
	EdgesAdded        int     `json:"edges_added"`
	PartialMstWeight  int     `json:"partial_mst_weight"`
}

type OutputJSON struct {
	InputFile        string            `json:"input_file"`
	FileReadTime     float64           `json:"file_read_time"`
	TotalTime        float64           `json:"total_time"`
	TotalIterations  int               `json:"total_iterations"`
	TotalMapTime     float64           `json:"total_map_time"`
	TotalReduceTime  float64           `json:"total_reduce_time"`
	MstWeight        int               `json:"mst_weight"`
	FinalComponents  int               `json:"final_components"`
	IterationDetails []IterationDetail `json:"iteration_details"`
}

type Graph struct {
	m_v         int
	m_edges     [][3]int
	m_component map[int]int
}

type edge struct {
	v, w int
}

func (g *Graph) findComponent(u int) int {
	if g.m_component[u] == u {
		return u
	}
	g.m_component[u] = g.findComponent(g.m_component[u])
	return g.m_component[u]
}

func (g *Graph) setComponent(u int) {
	if g.m_component[u] == u {
		return
	}
	for k := range g.m_component {
		g.m_component[k] = g.findComponent(k)
	}
}

func (g *Graph) union(componentSize []int, u, v int) {
	rootU := g.findComponent(u)
	rootV := g.findComponent(v)

	if rootU == rootV {
		return
	}

	if componentSize[rootU] <= componentSize[rootV] {
		g.m_component[rootU] = rootV
		componentSize[rootV] += componentSize[rootU]
	} else {
		g.m_component[rootV] = rootU
		componentSize[rootU] += componentSize[rootV]
	}
}

func (g *Graph) boruvka(inputFile string) OutputJSON {
	totalStartTime := time.Now()

	outputJSON := OutputJSON{
		InputFile:        inputFile,
		IterationDetails: []IterationDetail{},
	}

	startFileRead := time.Now()
	componentSize := make([]int, g.m_v)
	var mstWeight int
	outputJSON.FileReadTime = time.Since(startFileRead).Seconds()

	minimumWeightEdge := make([][]int, g.m_v)
	for i := range minimumWeightEdge {
		minimumWeightEdge[i] = []int{-1, -1, -1}
	}

	for node := 0; node < g.m_v; node++ {
		g.m_component[node] = node
		componentSize[node] = 1
	}

	numOfComponents := g.m_v
	iteration := 1

	totalMapTime := 0.0
	totalReduceTime := 0.0

	for numOfComponents > 1 {
		startMapTime := time.Now()
		preComponentCount := numOfComponents
		edgesAdded := 0

		for _, edge := range g.m_edges {
			u, v, w := edge[0], edge[1], edge[2]
			uComponent := g.findComponent(u)
			vComponent := g.findComponent(v)

			if uComponent != vComponent {
				if minimumWeightEdge[uComponent][0] == -1 || minimumWeightEdge[uComponent][2] > w {
					minimumWeightEdge[uComponent] = []int{u, v, w}
				}
				if minimumWeightEdge[vComponent][0] == -1 || minimumWeightEdge[vComponent][2] > w {
					minimumWeightEdge[vComponent] = []int{u, v, w}
				}
			}
		}
		mapTime := time.Since(startMapTime).Seconds()
		totalMapTime += mapTime

		startReduceTime := time.Now()
		for node := 0; node < g.m_v; node++ {
			if minimumWeightEdge[node][0] != -1 {
				u, v, w := minimumWeightEdge[node][0], minimumWeightEdge[node][1], minimumWeightEdge[node][2]
				uComponent := g.findComponent(u)
				vComponent := g.findComponent(v)

				if uComponent != vComponent {
					mstWeight += w
					g.union(componentSize, uComponent, vComponent)
					numOfComponents--
					edgesAdded++
				}
			}
		}
		reduceTime := time.Since(startReduceTime).Seconds()
		totalReduceTime += reduceTime

		outputJSON.IterationDetails = append(outputJSON.IterationDetails, IterationDetail{
			IterationNumber:   iteration,
			MapTime:           mapTime,
			ReduceTime:        reduceTime,
			PreComponentCount: preComponentCount,
			EdgesAdded:        edgesAdded,
			PartialMstWeight:  mstWeight,
		})

		for i := range minimumWeightEdge {
			minimumWeightEdge[i] = []int{-1, -1, -1}
		}
		if(edgesAdded == 0) {
			break
		}
		iteration++
	}

	totalTime := time.Since(totalStartTime).Seconds()

	outputJSON.TotalTime = totalTime
	outputJSON.TotalIterations = iteration - 1
	outputJSON.TotalMapTime = totalMapTime
	outputJSON.TotalReduceTime = totalReduceTime
	outputJSON.MstWeight = mstWeight
	outputJSON.FinalComponents = numOfComponents

	fmt.Println("Serial MST weight:", mstWeight)

	return outputJSON
}

func readGraph(filePath string) map[int][]edge {
	adjacencyList := make(map[int][]edge)

	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	
	// Skip header lines
	for scanner.Scan() {
		line := scanner.Text()
		if !strings.HasPrefix(line, "%") {
			break
		}
	}

	// Read matrix data
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())
		if len(parts) < 2 {
			continue
		}

		u, err1 := strconv.Atoi(parts[0])
		v, err2 := strconv.Atoi(parts[1])
		
		if err1 != nil || err2 != nil {
			continue
		}

		w := 1
		if len(parts) > 2 {
			w, _ = strconv.Atoi(parts[2])
		}

		// Ensure non-negative weight
		if w < 0 {
			w = -w
		}

		adjacencyList[u] = append(adjacencyList[u], edge{v, w})
		adjacencyList[v] = append(adjacencyList[v], edge{u, w})
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	return adjacencyList
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run main.go <mtx_file>")
		return
	}

	filePath := os.Args[1]
	adjacencyList := readGraph(filePath)

	graph := &Graph{
		m_v:         len(adjacencyList),
		m_component: make(map[int]int),
	}

	for u, neighbors := range adjacencyList {
		for _, neighbor := range neighbors {
			graph.m_edges = append(graph.m_edges, [3]int{u, neighbor.v, neighbor.w})
		}
	}

	output := graph.boruvka(filePath)

	outputFile := strings.Replace(filePath, ".mtx", "_serial.json", 1)
	jsonData, err := json.MarshalIndent(output, "", "  ")
	if err != nil {
		fmt.Println("Error marshaling JSON:", err)
		return
	}

	err = os.WriteFile(outputFile, jsonData, 0644)
	if err != nil {
		fmt.Println("Error writing JSON file:", err)
		return
	}

	fmt.Println("Output JSON file created:", outputFile)
}